src/de/raw.rs

```diff
@@ -0,0 +1,947 @@
+                .deserialize(FieldDeserializer {
+                    field_name: "base64",
+                })
+                .map(Some),
+            BinaryDeserializationStage::Done => Ok(None),
+        }
+    }
+
+    fn next_value_seed<V>(&mut self, seed: V) -> Result<V::Value>
+    where
+        V: serde::de::DeserializeSeed<'de>,
+    {
+        seed.deserialize(&mut *self.deserializer)
+    }
+}
+
+struct BinaryDeserializer {
+    binary: Binary,
+    stage: BinaryDeserializationStage,
+}
+
+impl BinaryDeserializer {
+    fn new(binary: Binary) -> Self {
+        Self {
+            binary,
+            stage: BinaryDeserializationStage::TopLevel,
+        }
+    }
+}
+
+impl<'de, 'a> serde::de::Deserializer<'de> for &'a mut BinaryDeserializer {
+    type Error = Error;
+
+    fn deserialize_any<V>(mut self, visitor: V) -> Result<V::Value>
+    where
+        V: serde::de::Visitor<'de>,
+    {
+        match self.stage {
+            BinaryDeserializationStage::TopLevel => {
+                self.stage = BinaryDeserializationStage::Subtype;
+                visitor.visit_map(BinaryAccess {
+                    deserializer: &mut self,
+                })
+            }
+            BinaryDeserializationStage::Subtype => {
+                self.stage = BinaryDeserializationStage::Bytes;
+                visitor.visit_string(hex::encode([u8::from(self.binary.subtype)]))
+            }
+            BinaryDeserializationStage::Bytes => {
+                self.stage = BinaryDeserializationStage::Done;
+                visitor.visit_string(base64::encode(self.binary.bytes.as_slice()))
+            }
+            BinaryDeserializationStage::Done => {
+                Err(Error::custom("Binary fully deserialized already"))
+            }
+        }
+    }
+
+    serde::forward_to_deserialize_any! {
+        bool u8 u16 u32 u64 i8 i16 i32 i64 f32 f64 char str string seq
+        bytes byte_buf map struct option unit newtype_struct
+        ignored_any unit_struct tuple_struct tuple enum identifier
+    }
+}
+
+enum BinaryDeserializationStage {
+    TopLevel,
+    Subtype,
+    Bytes,
+    Done,
+}
+
+/// Struct wrapping a slice of BSON bytes.
+struct BsonBuf<'a> {
+    bytes: &'a [u8],
+    index: usize,
+}
+
+impl<'a> Read for BsonBuf<'a> {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+        self.index_check()?;
+        let bytes_read = self.bytes[self.index..].as_ref().read(buf)?;
+        self.index += bytes_read;
+        Ok(bytes_read)
+    }
+}
+
+impl<'a> BsonBuf<'a> {
+    fn new(bytes: &'a [u8]) -> Self {
+        Self { bytes, index: 0 }
+    }
+
+    fn bytes_read(&self) -> usize {
+        self.index
+    }
+
+    /// Verify the index has not run out of bounds.
+    fn index_check(&self) -> std::io::Result<()> {
+        if self.index >= self.bytes.len() {
+            return Err(ErrorKind::UnexpectedEof.into());
+        }
+        Ok(())
+    }
+
+    fn read_cstr(&mut self) -> Result<&'a str> {
```

Comment:
Since the common case for invalid UTF-8 was server error message values rather than in the keys, I opted to only do the lossy handling for those. That being said, it probably makes more sense to be consistent throughout. Once we decide a way forward on the lossy UTF-8 situation, I'll unify this with `read_str`.